nbs_generate_schematic.py

```diff
@@ -4,221 +4,239 @@
-    currentModule = 1
-    for module in contents:
-      lowerChest1 = ''
-      upperChest1 = ''
-      lowerChest2 = ''
-      upperChest2 = ''
-      lowerShulker = ''
-      upperShulker = ''
-      currentShulker = 1
-      lowerOctaveEmpty = len(module[0]) == 0
-      upperOctaveEmpty = len(module[1]) == 0
-      for currentTick in range(songLengthAdjusted):
-        currentSlot = currentTick % 27
-        if lowerOctaveEmpty == 0:
-          lowerShulker += newDisc(currentSlot, module[0][currentTick]) + ','
-        if upperOctaveEmpty == 0:
-          upperShulker += newDisc(currentSlot, module[1][currentTick]) + ','
-        # if we are on the last slot of a shulker box, or the song has ended
-        if (currentTick + 1) % 27 == 0 or currentTick == songLengthAdjusted - 1:
-          # turn the shulker contents into actual shulker
-          if lowerOctaveEmpty == 0:
-            lowerShulker = createShulker(currentShulker, lowerShulker)
-          if upperOctaveEmpty == 0:
-            upperShulker = createShulker(currentShulker, upperShulker)
-          # if the current shulker should go in the first chests
-          if currentShulker <= 27:
-            if lowerOctaveEmpty == 0:
-              lowerChest1 += lowerShulker + ','
-            if upperOctaveEmpty == 0:
-              upperChest1 += upperShulker + ','
-          else:
+        allChestContents[instrument] = []
+        for i in range(CHORD_MAX_SIZES[instrument]):
+            allChestContents[instrument].append([emptyChest.copy(), emptyChest.copy()])
+
+    # iterate through the whole song by chords
+    keyModifier = INSTRUMENT_RANGE[0]
+    currentIndices = {}
+    for tick, chord in song:
+        # reset current indices
+        for instrument in INSTRUMENTS:
+            currentIndices[instrument] = [0, 0]
+
+        for note in chord:
+            instrument = INSTRUMENTS[note.instrument]
+            adjustedKey = note.key - keyModifier
+            octave = 0 if adjustedKey <= 11 else 1
+            allChestContents[instrument][currentIndices[instrument][octave]][octave][tick] = adjustedKey
+            currentIndices[instrument][octave] += 1
+
+    minimalChestContents = removeEmptyChests(allChestContents)
+
+    # turn minimalChestContents into a schematic
+    schem = mcschematic.MCSchematic()
+    offset = 0
+    print('Generating Schematic...')
+    for instrument, contents in minimalChestContents.items():
+        currentModule = 1
+        for module in contents:
+            lowerChest1 = ''
+            upperChest1 = ''
+            lowerChest2 = ''
+            upperChest2 = ''
+            lowerShulker = ''
+            upperShulker = ''
+            currentShulker = 1
+            lowerOctaveEmpty = len(module[0]) == 0
+            upperOctaveEmpty = len(module[1]) == 0
+            for currentTick in range(songLengthAdjusted):
+                currentSlot = currentTick % 27
+                if lowerOctaveEmpty == 0:
```

Comment:
Since we are changing to booleans, I believe that all of these if statements can be changed to reflect that.